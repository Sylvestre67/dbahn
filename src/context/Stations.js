import React from 'react';
// import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { withRouter } from 'react-router-dom';

import apolloClient from '../services/apollo';

import { Query } from 'react-apollo';

const GET_STATIONS = gql`
	query stations($search: String!) {
		search(searchTerm: $search) {
			stations {
				primaryEvaId
				name
				picture {
					url
				}
			}
		}
	}
`;

export const StationsContext = React.createContext({});

class Stations extends React.PureComponent {
	constructor(props) {
		super(props);

		this.state = {
			data: [],
			search: '',
		};
	}

	componentDidMount() {
		const { location } = this.props;
		const { pathname, search } = location;

		if (pathname === '/' && search.replace('?q=', '').length) {
			this.setState({ search: search.replace('?q=', '') });
		}
	}

	componentDidUpdate(prevProps, prevState) {
		const { location } = this.props;
		const { pathname, key, search } = location;

		if (pathname === '/' && key !== prevProps.location.key) {
			this.setState({ search: search.replace('?q=', '') });
		}
	}

	render() {
		const { search } = this.state;

		return (
			<Query
				query={GET_STATIONS}
				variables={{ search }}
				client={apolloClient}>
				{({ data = { search: { stations: [] } }, loading, error }) => (
					<StationsContext.Provider
						value={{
							data: data.search ? data.search.stations : [],
							loading,
							error,
						}}>
						{this.props.children}
					</StationsContext.Provider>
				)}
			</Query>
		);
	}
}

export default withRouter(Stations);
