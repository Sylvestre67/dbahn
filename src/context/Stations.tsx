import React from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';

import apolloClient from '../services/apollo';
import { StationDefs } from './__generated__/StationDefs';

import StationsQuery from './StationsQuery';

import { SEARCH_STATIONS } from './queries';
import { ApolloError } from 'apollo-client';

export interface IStationsProps extends RouteComponentProps {}

export interface IStationsState {
	data?: StationDefs;
	search: string;
}

export interface IStationsContext {
	data: StationDefs | undefined;
	loading: boolean;
	error: ApolloError | undefined;
}

export const StationsContext = React.createContext<IStationsContext | null>(
	null
);

class Stations extends React.PureComponent<IStationsProps, IStationsState> {
	constructor(props: IStationsProps) {
		super(props);

		this.state = {
			search: '',
		};
	}

	componentDidMount() {
		const { location } = this.props;
		const { pathname, search } = location;

		if (pathname === '/' && search.replace('?q=', '').length) {
			this.setState({ search: search.replace('?q=', '') });
		}
	}

	componentDidUpdate(prevProps: IStationsProps) {
		const { location } = this.props;
		const { pathname, key, search } = location;

		if (pathname === '/' && key !== prevProps.location.key) {
			this.setState({ search: search.replace('?q=', '') });
		}
	}

	render() {
		const { search } = this.state;

		if (search.length) {
			return (
				<StationsQuery
					query={SEARCH_STATIONS}
					variables={{ search }}
					client={apolloClient}>
					{({ data, loading, error }) => (
						<StationsContext.Provider
							value={{
								data,
								loading,
								error,
							}}>
							{this.props.children}
						</StationsContext.Provider>
					)}
				</StationsQuery>
			);
		}

		return <React.Fragment>{this.props.children}</React.Fragment>;
	}
}

export default withRouter(Stations);
