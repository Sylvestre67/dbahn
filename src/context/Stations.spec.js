import React from 'react';
import { shallow } from 'enzyme';

import Stations from './Stations';
let component;

const props = {};

describe('Stations', () => {
	beforeAll(() => {
		component = shallow(<Stations {...props} />);
	});

	it('should render correctly', () => {
		expect(component).toMatchSnapshot();
	});
});
