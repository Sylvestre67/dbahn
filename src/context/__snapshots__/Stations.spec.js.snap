// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stations should render correctly 1`] = `
<Route>
  <Component />
</Route>
`;
